harfbuzz_overflow_resolution.md

```diff
@@ -0,0 +1,157 @@
+# Introduction
+
+Several tables in the opentype format are formed internally by a graph of subtables. Parent node's
+reference their children through the use of positive offsets, which are typically 16 bits wide.
+For storage in the font file the graph must be given a topological ordering and then the subtables
+packed in serial according to that ordering. Since 16 bit offsets have a maximum value of 65,535 if
+the distance between a parent subtable and a child is more then 65,535 bytes then it's not possible
+for the offset to encode that edge.
+
+For many fonts with complex layouts it's not unusual for the GSUB/GPOS table to be larger than 65kb.
+As a result these types of fonts are susceptible to offset overflows when serializing to the binary
+font format.
+
+Offset overflows can be resolved through a variety of strategies:
+*  Often times a different topological ordering can avoid the overflows.
+*  In the graph subtables can have many parents. Which can result in the link from furthest parent(s)
+   being at risk for overflows. In these cases it's possible to duplicate the shared subtable which
+   allows it to be placed closer to it's parent.
+*  If subtables exist which are themselves larger than 65kb it's not possible for any offsets to point
+   past them. In these cases the subtable can usually be split into two smaller subtables to allow
+   for more flexibility in the ordering.
+*  In GSUB/GPOS overflows from Lookup subtables can be resolved by changing the Lookup to an extension
+   lookup which uses a 32 bit offset instead of 16 bit offset.
+   
+In general there isn't a simple solution to produce an optimal topological ordering for a given graph.
+Finding an ordering which doesn't overflow is a NP hard problem. Existing solutions use heuristics
+which attempt a combination of the above strategies to attempt to find a non-overflowing configuration.
+   
+The harfbuzz subsetting library
+[includes a repacking algorithm](https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-repacker.hh)
+which is used to resolve offset overflows that are present in the subsetted tables it produces. This
+document provides a deep dive into how the harfbuzz repacking algorithm works.
+
+# Foundations
+
+There's three key pieces to the harfbuzz approach:
+
+*  Graph representation: a table's internal structure is represented by a lightweight graph
+   representation where each subtable is a node and each offset forms an edge. The actual subtable
+   bytes are not stored in each node. Instead we just store start and end memory addresses for each.
+   This lightweight representation can be quickly modified as the repacking algorithm iterates.
```

Comment:
How the graph is represented is interesting but sort of seems like an implementation detail, not a fundamental characteristic of the approach?

Reply:
I'll try and reword this, but the important part is we're not modifying the bytes of the subtables until we're satisified we've solved the overflow. This is in contrast to fontTools implementation which repeatedly tries to recompile the table (generating the serialized subtable bytes until it hits an overflow) on every iteration.

Reply:
> the important part is we're not modifying the bytes of the subtables until we're satisified we've solved the overflow

Could probably say that somewhat more directly? And then explain how (which I think is what it has now)